compiler/rustc_lint/src/internal.rs

```diff
@@ -364,54 +367,133 @@ declare_lint_pass!(Diagnostics => [UNTRANSLATABLE_DIAGNOSTIC, DIAGNOSTIC_OUTSIDE
+
+        // Calls to `#[rustc_lint_diagnostics]`-marked functions should only occur:
+        // - inside an impl of `IntoDiagnostic`, `AddToDiagnostic`, or `DecorateLint`, or
+        // - inside a parent function that is itself marked with `#[rustc_lint_diagnostics]`.
+        //
+        // Otherwise, emit a `DIAGNOSTIC_OUTSIDE_OF_IMPL` lint.
+        if has_attr && !parent_has_attr {
+            let mut is_inside_appropriate_impl = false;
+            for (_hir_id, parent) in cx.tcx.hir().parent_iter(expr.hir_id) {
+                debug!(?parent);
+                if let Node::Item(Item { kind: ItemKind::Impl(impl_), .. }) = parent
+                    && let Impl { of_trait: Some(of_trait), .. } = impl_
+                    && let Some(def_id) = of_trait.trait_def_id()
+                    && let Some(name) = cx.tcx.get_diagnostic_name(def_id)
+                    && matches!(
+                        name,
+                        sym::IntoDiagnostic | sym::AddToDiagnostic | sym::DecorateLint
+                    )
+                {
+                    is_inside_appropriate_impl = true;
+                    break;
+                }
+            }
+            debug!(?is_inside_appropriate_impl);
+            if !is_inside_appropriate_impl {
+                cx.emit_span_lint(DIAGNOSTIC_OUTSIDE_OF_IMPL, span, DiagOutOfImpl);
+            }
+        }
+
+        // Calls to methods with an `impl Into<{D,Subd}iagnosticMessage>` parameter must be passed
+        // an arg with type `{D,Subd}iagnosticMessage` or `impl Into<{D,Subd}iagnosticMessage`.
+        // Otherwise, emit an `UNTRANSLATABLE_DIAGNOSTIC` lint.
+        //
+        // This is imprecise. It doesn't match args to parameters exactly, and could mis-analyze a
+        // function like `fn foo(msg1: DiagnosticMessage, msg2: impl Into<DiagnosticMessage>)`, but
+        // there are no such functions in practice, so this is good enough for now.
+        if num_impl_into_diagnostic_message_params > 1 {
+            cx.tcx.dcx().span_bug(span, "multiple `impl Into<{D,Sub}diagnosticMessage>` params")
+        } else if num_impl_into_diagnostic_message_params == 1 {
+            let mut has_diagnostic_message_arg = false;
+            let mut has_impl_into_diagnostic_message_arg = false;
+            for ty in call_tys {
+                // Is the argument type `{Sub,D}iagnosticMessage`?
+                if let Some(adt_def) = ty.ty_adt_def()
+                    && let Some(name) = cx.tcx.get_diagnostic_name(adt_def.did())
+                    && matches!(name, sym::DiagnosticMessage | sym::SubdiagnosticMessage)
+                {
+                    has_diagnostic_message_arg = true;
+                    break;
+                }
+
+                // Is the argument type `impl Into<{Sub,D}iagnosticMessage>`?
+                if let ty::Param(param_ty) = ty.kind()
+                    && let name = param_ty.name.as_str()
```

Comment:
I think these, uh, checks can just be left out? If someone passes a ty::Param to an Into<DiagnosticMessage> parameter, then the type parameter has to have bounds on these traits.

Reply:
As per my comment below, the code higher up determines that there is an `impl Into<DiagnosticMessage>` parameter (but isn't sure which ones). Then this code checks all the arguments for one with a type that doesn't require translation. E.g. `String` requires translation, but `DiagnosticMessage` doesn't. Also, an argument of type `impl Into<DiagnosticMessage>` doesn't need translation.

If I remove this entire `if let` (not sure if that's what you mean by "these checks" then I get false positives in methods like this:
```rust
    pub fn buffer_lint(
        &mut self,
        lint: &'static Lint,
        id: NodeId,
        sp: impl Into<MultiSpan>,
        msg: impl Into<DiagnosticMessage>,
    ) {
        self.add_lint(lint, id, sp.into(), msg, BuiltinLintDiagnostics::Normal)
    }
```
where `msg`, which has type `impl Into<DiagnosticMessage>` gets passed along untouched.